src/Domain/DonationRepository.php

```diff
@@ -257,8 +265,6 @@ public function releaseMatchFunds(Donation $donation): void
 
         $this->logInfo("Taking from ID {$donation->getUuid()} released match funds totalling {$totalAmountReleased}");
         $this->logInfo('Deallocation took ' . round($lockEndTime - $lockStartTime, 6) . ' seconds');
-
-        $fundsReleaseLock->release();
```

Comment:
Sorry, I should have copied more context to here.

I don't think we can deal with it that cleanly because two separate processes are involved. From my initial investigations on Slack:

> Right, so looks like a very unlucky situation where the natural expiry command and an explicit Cancel from the browser (e.g. from changing amount) led to the same release of match funds within 0.5 seconds.

> Kind of a tricky one to fully preempt because we almost certainly want to keep releasing funds immediately on Cancel. Even waiting a minute means someone changing amounts might be told match funds are unavailable, unnecessarily.
> The challenge comes because there is a decent chance that these steps in order lead to two processes wanting to release match funds:
> 1. regular cron reads donations ready to release funds and finds Collected donation X (database query and looping through takes up to a few seconds)
> 2. during the few seconds, donor cancels donation X
> 3. web process doesn’t know about cron so immediately releases funds, finishes its work and releases the lock
> 4. cron has the previous info in memory as it works through the loop and so also still wants to release funds, and is able to acquire a new lock and do so
>
> It seems like our best option is probably some kind of longer lasting lock. Certainly for when things happen this way around, maybe both ways around? If it’s the latter then it could be that we just ought to rely on the TTL and not release the lock explicitly after any release of match funds; perhaps making an exception if we need to newly allocate match funds to an existing donation a second time.

Reply:
@NoelLH Thanks. Lets have a call after standup. That helps but I'm still not fully clear.

It makes me think maybe what we need then is for the cron process to hold the lock for basically the entire time `MatchFundsRedistributor::redistributeMatchFunds` is running, from before it fetches data from the DB until its done. Or maybe better split that function up so that it queries for a list of donation IDs not full donation objects, then for each one acquires the lock, loads it from db, then releases the lock.

Also probably longer term thing but maybe instead of just sending a cancel command the client should send a command to replace one donation with another (effectively changing the amount on a donation) so that matchbot can do the entire thing atomically. That's what it looks like in the UI anyway.

Reply:
Also wondering if there's something we can do to stop using or de-emphasize the matchbot.CampaignFunding.amountAvailable column . As I understand the issue is overlapping processes to update this column that can mean it ends up with the wrong value. Maybe instead of storing the amountAvailable in MySQL we can store the full list of things that increase and decrease that amount, and calculate the current total on demand. As that list would only ever need to appended to it wouldn't need locking. If necessary we could also keep the amountAvailable as a performance optimisation but the full list would be the main source of truth.

Like my bank account has a balance but the source of truth is the list of transactions from which the balance can be calculated, not the balance itself.

Reply:
Clarified with @NoelLH , the thing we're waiting for before releasing the lock is any copy of the donation thats in memory on a different thread from before this lock was aquired needs to be released. As threads do not run for more than 5 minutes we can be fairly sure that that will happen if we just wait 5 minutes.